connector/connect/common/src/main/protobuf/spark/connect/relations.proto

```diff
@@ -353,11 +353,16 @@ message LocalRelation {
   optional bytes data = 1;
 
   // (Optional) The schema of local data.
-  // It should be either a DDL-formatted type string or a JSON string.
   //
   // The server side will update the column names and data types according to this schema.
   // If the 'data' is not provided, then this schema will be required.
-  optional string schema = 2;
+  oneof schema {
```

Comment:
since we had introduced `DDLParse` in https://github.com/apache/spark/commit/0c19ea4773b8eaeedc4031cb1c5c3e97f2c2d3b9 

what about always parsing the ddl string in client and only use `DataType dataType` in protos? I feel it will make client simpler.

Reply:
My motivation was to avoid roundtrips, but let's discuss.

Reply:
in Python Client, there are a few edge cases in `createDataFrame`:

1, `spark.createDataFrame([(100, None)], "age INT, name STRING")` [SPARK-42458](https://issues.apache.org/jira/browse/SPARK-42458)
can not infer the schema since the second column only contains `None`, https://github.com/apache/spark/pull/40240 fixed by parse the ddl string via `self.createDataFrame([], schema=_schema_str).schema`;

2, `self.spark.createDataFrame([["a", "b"]], ["col1"])` [SPARK-42022](https://issues.apache.org/jira/browse/SPARK-42022)
the `createDataFrame` should be able to autogenerate missing column names, if we don't do this in client, we may need to do it in `LocalRelation`

There are still some other behavior differences, If we can parse the DDL string in client, I guess we can reuse more functions in PySpark's `createDataFrame`.

cc @HyukjinKwon @ueshin

Reply:
In some cases, we need to parse it beforehand anyway, for the case of #40240, we can't create converter from the local python object to Arrow table without the schema as `DataType` object.

Another example is UDFs. It needs to pickle the function, and its return type as `DataType` object. We can't generate Python `DataType` object and pickle it in the `command` field in server side anymore. So we need to parse the DDL string beforehand.

Reply:
I think this change is fine. I'd rather like to change all places to accept both string and `DataType` and the choice is up to the client developers.

Reply:
@ueshin why not implement a parser on the python side? That shouldn't be too hard, and it saves you from doing a lot of expensive rpcs.

Reply:
Good point. IIRC, there was a decision not to implement it in Python.
cc @HyukjinKwon @zhengruifeng

Reply:
> I'd rather like to change all places to accept both string and DataType and the choice is up to the client developers.

I think it is a good idea. We may add a dedicated message for it and then use it in all places.

Are we going to do this?

also cc @LuciferYang

Reply:
Thanks @zhengruifeng

Reply:
On second thoughts, if `UnparsedDataType` at #40260 is good to go, we might not need to use DDL string anymore, and just use `DataType`.